src/input/shared.rs

```diff
@@ -109,7 +109,7 @@ pub fn str_as_float<'s>(input: &'s impl Input<'s>, str: &str) -> ValResult<Eithe
 
 /// parse a string as an int, `input` is required here to get lifetimes to match up
 ///
-fn _parse_str<'s>(_input: &'s impl Input<'s>, str: &str, len: usize) -> Option<EitherInt<'s>> {
+fn _parse_str<'py>(_input: &(impl Input<'py> + ?Sized), str: &str, len: usize) -> Option<EitherInt<'py>> {
```

Comment:
There's a lot of `&(impl Input<'py> + ?Sized),` can you explain better what it does/why we're using it?